sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1194,6 +1195,41 @@ object RemoveDispensableExpressions extends Rule[LogicalPlan] {
 }
 
 /**
+ * Removes useless Distinct that are not necessary.
+ */
+object EliminateDistinct extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    // Eliminate the useless distinct.
+    // Distinct has been replaced by Aggregate in the rule ReplaceDistinctWithAggregate
+    case a @ Aggregate(grouping, aggs, child) if isDistinct(a) && isDistinct(child) => child
+  }
+
+  // propagate the distinct property from the child
+  @tailrec
```

Comment:
Another solution is to add a property `isDistinct` to `LogicalPlan`. However, it could be expensive for recursive calls, compared with the `@tailrec`. In the future, if the physical plan will use the property `isDistinct`, we can rewrite it. Actually, this is a very critical property in runtime. Thanks!